aiokafka/fetcher.py

```diff
@@ -0,0 +1,579 @@
+                              " since it is no longer assigned", tp)
+                    continue
+
+                # note that the consumed position should always be available
+                # as long as the partition is still assigned
+                position = self._subscriptions.assignment[tp].position
+                if not self._subscriptions.is_fetchable(tp):
+                    # this can happen when a partition consumption paused
+                    # before fetched records are returned
+                    log.debug(
+                        "Not returning fetched records for assigned partition"
+                        " %s since it is no longer fetchable", tp)
+
+                elif fetch_offset == position:
+                    log.debug(
+                        "Returning fetched records at offset %d for assigned"
+                        " partition %s", position, tp)
+                    for msg in self._unpack_message_set(tp, messages):
+                        # Because we are in a generator, it is possible for
+                        # subscription state to change between yield calls
+                        # so we need to re-check on each loop
+                        # this should catch assignment changes, pauses
+                        # and resets via seek_to_beginning / seek_to_end
+                        if tp not in self._subscriptions.assignment:
+                            log.debug("Not returning fetched records"
+                                      " bcs partitions are unassigned")
+                            break
+                        position = self._subscriptions.assignment[tp].position
+                        if not self._subscriptions.is_fetchable(tp):
+                            log.debug(
+                                "Not returning fetched records for partition"
+                                " %s since it is no longer fetchable", tp)
+                            break
+
+                        if self._subscriptions.needs_partition_assignment:
+                            log.debug("Not returning fetched records"
+                                      " bcs partitions are unassigned")
+                            break
+
+                        # Compressed messagesets may include earlier messages
+                        # It is also possible that the user called seek()
+                        elif msg.offset != position:
+                            continue
+
+                        self._subscriptions.assignment[tp].position += 1
+                        yield msg
+                else:
+                    # these records aren't next in line based on the last
+                    # consumed position, ignore them they must be from
+                    # an obsolete request
+                    log.debug("Ignoring fetched records for %s at offset %s",
+                              tp, fetch_offset)
+
+    def next_record(self, partitions):
```

Comment:
This needs some kind of traversing algorithm. Currently it's implemented in the iterator, but I think it's behaviour is not obvious (for example what will happen if we try to add an element to `tp_records` while it's in process of iteration?). 

What is needed is that we'd always process partitions equaly. How about:
* We maintain an ordered dict instead of a simple dict?
* Set a waiting future if no records prefetched. It may be released at each successful fetch request (if we waited for other partitions - just set another waiter).